### DISCUSSION TITLE
Combination of `x-bind` and `$refs` not working (since Alpine.js 3.x)

### DISCUSSION THREAD

**Discussion by tobihagemann** [alpinejs/alpine#1863] [7. Bugs (archived)] [answered]
Not sure if my title even describes the bug properly, so let me give you an example:

This doesn't work (with Alpine.js 3.2.2): https://jsfiddle.net/2Lrfoexg/
This works (with Alpine.js 2.8.2): https://jsfiddle.net/9o817fyx/

What I'm trying to achieve: Enable/disable a button based on `$refs.form.checkValidity()`. In the example, you have to enter a "valid" email address for the button to become enabled.

**Comment by KevinBatdorf** [comment]
In V2 the `$refs` were queries when invoked, where now they are cached at runtime. You're better off adding your own helper or just using standard JS to grab the element. 

There are multiple ways to make a helper for this. Here's one I was messing around with a bit ago: https://codepen.io/KevinBatdorf/pen/VwbvZBR?editors=1010

Probably better ways to get it working though specific to your use case.

**Reply by tobihagemann** [reply]
I couldn't get your solution to work and I also didn't get a solution via "using standard JS to grab the element" to work. Is it just not working inside `x-bind`? I'll do more experiments but I'm now really confused. :smile:

**Comment by SimoTod** [comment, marked as answer]
The issue is not the one suggested. Basically in V3, the reactivity works in a different way. 
v2 updates the whole component at any change
v3 only updates directives that use "reactive" data that has changed. Note the reactive bit here is important, it means (to simplify) anything that is in the x-data attribute.

Your form, to work properly, should be something like
```html
<form x-data="{email: ''}" x-ref="form">
  <input x-model="email" type="email" required/>
  <button type="submit" @click="$event.preventDefault()" :disabled="email && !$refs.form.checkValidity()">Submit</button>
</form>
```
In this case, `:disabled` will be reevaluated every time email changes.

It can quickly become a nightmare if you have multiple fields in your form though, in that case you probably want to think about a different solution.

**Reply by tobihagemann** [reply]
Ah, I suspected that and probably tried the wrong thing but that definitely helped, thank you! Indeed, multiple fields would become a problem that I have to think about. :sweat_smile: I wasn't sure if something like `x-effect` would help and then set `disabled` "from the outside" (without `x-bind`). Will have to give it a try.

Edit: Or not... overlooked the last part [in the documentation](https://alpinejs.dev/directives/effect).

> You can think of it as a watcher where you don't have to specify what property to watch, it will watch all properties **used within it**.

**Reply by tobihagemann** [reply]
Is it worth opening a feature request for something like `$watchAll` or `x-effect-all`? I mean... I don't really see a way to watch all properties with the current feature set.

Edit: Or I could write my own magic/directive. But... I honestly don't know how.
